packages/squiggle-lang/src/rescript/FR/FR_Plot.res

```diff
@@ -0,0 +1,148 @@
+    }
+
+    let merge = (m1: t<'a>, m2: t<'b>): t<('a, 'b)> => {
+      {
+        result: (a: Reducer_T.map) => E.R.merge(m1.result(a), m2.result(a)),
+        typesRequired: Belt.Array.concat(m1.typesRequired, m2.typesRequired),
+      }
+    }
+
+    let fmap = (f: 'a => 'b, m: t<'a>): t<'b> => {
+      {
+        result: (a: Reducer_T.map) => E.R.fmap(f, m.result(a)),
+        typesRequired: m.typesRequired,
+      }
+    }
+
+    let app = (m1: t<'a => 'b>, m2: t<'a>): t<'b> => {
+      {
+        result: (a: Reducer_T.map) =>
+          E.R.merge(m1.result(a), m2.result(a))->E.R2.fmap(((f, x)) => f(x)),
+        typesRequired: Belt.Array.concat(m1.typesRequired, m2.typesRequired),
+      }
+    }
+  }
+
+  let getString: fnApp<string> = {
+    let func = (a: Reducer_T.value) =>
+      switch a {
+      | IEvString(s) => Ok(s)
+      | _ => Error(impossibleError)
+      }
+    {result: func, typeRequired: FRTypeString}
+  }
+
+  let getArray = (child: fnApp<'a>): fnApp<array<'a>> => {
+    let func = (a: Reducer_T.value) =>
+      switch a {
+      | IEvArray(x) => x->E.A2.fmap(child.result)->E.A.R.firstErrorOrOpen
+      | _ => Error(impossibleError)
+      }
+    {result: func, typeRequired: FRTypeArray(child.typeRequired)}
+  }
+  let getRecord = (recMonad: Record.t<'a>): fnApp<'a> => {
+    let func = (a: Reducer_T.value) =>
+      switch a {
+      | IEvRecord(s) => recMonad.result(s)
+      | _ => Error(impossibleError)
+      }
+    {result: func, typeRequired: FRTypeRecord(recMonad.typesRequired)}
+  }
+
+  let getDistOrNumber: fnApp<GenericDist.t> = {
+    let func = (a: Reducer_T.value) =>
+      switch a {
+      | IEvDistribution(s) => Ok(s)
+      | IEvNumber(s) => Ok(GenericDist.fromFloat(s))
+      | _ => Error(impossibleError)
+      }
+    {result: func, typeRequired: FRTypeDistOrNumber}
+  }
+
+  let oneArgDef = (
+    name: string,
+    arg1: fnApp<'a>,
+    def: 'a => result<Reducer_T.value, SqError.Message.t>,
+  ): FnDefinition.t =>
+    FnDefinition.make(
+      ~name,
+      ~inputs=[arg1.typeRequired],
+      ~run=(inputs, _, _) => {
+        E.R.bind(arg1.result(inputs[0]), def)
+      },
+      (),
+    )
+}
+
+module Internals = {
+  let makeLabeledDistribution = (
+    name: string,
+    distribution: GenericDist.t,
+  ): Reducer_T.labeledDistribution => {name: name, distribution: distribution}
+
+  let getLabeledDistribution: FnApp.fnApp<Reducer_T.labeledDistribution> = {
+    makeLabeledDistribution
+    ->FnApp.Record.fmap(FnApp.Record.getField("title", FnApp.getString))
+    ->FnApp.Record.app(FnApp.Record.getField("value", FnApp.getDistOrNumber))
+    ->FnApp.getRecord
+  }
+
+  let makePlot = (show: array<Reducer_T.labeledDistribution>): Reducer_T.plotValue => {
+    distributions: show,
+  }
+
+  let parsePlotValue: FnApp.fnApp<Reducer_T.plotValue> = {
+    makePlot
+    ->FnApp.Record.fmap(FnApp.Record.getField("dists", FnApp.getArray(getLabeledDistribution)))
+    ->FnApp.getRecord
+  }
+}
```

Comment:
Thoughts:
- this is impressive and really clever
- but also complicated and I'm not sure if it's worth it — the value brought by the ability to track the output frType together with the result shape seems less to me than the complexity of the API
- I'd be more ok with it if the end user API was nicer; right now it's leaking the implementation details too much
- this makes me want to give up on Rescript in favor of Typescript, it really shouldn't be this hard

On that last point. Could we have something like this?
```
  let getLabeledDistribution = {
    T.record(
      (name, distribution) => { name, distribution }
      ->T.Record.field("title", T.string)
      ->T.Record.field("value", T.distOrNumber)
    )
  }
```

I _think_ the unification of `Record.app` and `Record.fmap` under a single `Record.field` call is doable (maybe with a wrapper over the initial record builder function that builds the initial `Record.t` value?). Other changes in this snippet should be possible.